### ISSUE TITLE
Add probe_PAYGAS_mode utility function

### ISSUE THREAD

**Issue by NIC619** [ethereum/py-evm#371]
### What is wrong?
After `gas_price` is abstracted away and `PAYGAS` is introduced, miner SHOULD(but not mandatory) check if `PAYGAS` is triggered before fully executing a transaction.

### How can it be fixed
Add a `probe_PAYGAS_mode` utility function which executes a transaction with a rather small amount of gas. If miner wants to make sure they will be paid, they should call this function before applying a transaction.

This function computationally normally execute a transaction. Normally this execution terminates with `OutOfGas` error. After the execution terminates, check if `PAYGAS` is triggered and return result.

There is one adjustable system parameter `probe_PAYGAS_mode_gas` that specifies how much gas is provisioned for the execution. Miner should be able to adjust this value depending on how much gas(computation resource) they are willing to risk.

Possible extensions:
- This function returns `True` of `False` indicating if miner will get paid. Additionally we can also return the value of gas price set by `PAYGAS`, to help prioritizing transactions in transaction pool.
- modify `PAYGAS`: If `PAYGAS` is triggered in `probe_PAYGAS_mode`, terminates immediately. This helps saving some computation otherwise it terminates only when gas has run out.
- Should probably check if it terminates in ways that we are expecting, e.g, `OutOfGas` or `PAYGAS` triggered.

**Comment by pipermerriam** [comment]
I had the following thought.  Modifying the gas limit could have the unfortunate side effect of invalidating certain transactions that would normally succeed *if* the contract code has any logic switches based on the amount of gas provided.

To fix this, we'd need to inject the *check* all the way down into the `Computation.apply_computation` method.  This has the side-benefit of not needing to double execute every transaction.

```python
# in `computation.py`
class PAYGASFriendlyComputationMixin:
    @classmethod
    def apply_computation(cls, vm_state, message):
        """
        Perform the computation that would be triggered by the VM message.
        """
        with cls(vm_state, message) as computation:
            # Early exit on pre-compiles
            if message.code_address in computation.precompiles:
                computation.precompiles[message.code_address](computation)
                return computation

            for opcode in computation.code:
                #
                # inject check for whether PAYGAS has been called and abort with special 
                # exception if it has not and we've exceeded the gas requirements.
                #
                opcode_fn = computation.get_opcode_fn(computation.opcodes, opcode)

                computation.logger.trace(
                    "OPCODE: 0x%x (%s) | pc: %s",
                    opcode,
                    opcode_fn.mnemonic,
                    max(0, computation.code.pc - 1),
                )

                try:
                    opcode_fn(computation=computation)
                except Halt:
                    break
        return computation

# in miner script
try:
    chain.apply_transaction(transaction)
except SomeCustomPAYGASError:
    # discard transaction and continue
    ...
```

**Comment by jannikluhn** [comment]
To make this more general, how about adding an optional `post_opcode_callback` (or `pre_opcode_callback`) parameter at the lowest level? The callback could inspect the state of the computation and, if necessary, raise an exception (or return `False`) to abort the execution. Miners could add some `check_gas_payment` as a callback, but this could also be used for other things (mainly debugging and testing, but maybe also as an easy to access but low level entry point for chain analysis or contract watcher tools). Disadvantage would be that the parameter would have to be added to the whole stack of `apply_computation` functions and friends, so it's not as non-invasive as the mixin solution.

**Comment by pipermerriam** [comment]
I think requiring it to be passed through the whole stack is a no-go for a callback based solution.  I have ruminated on solutions for more elegant modification of the `Chain` object and all of it's various dependencies.  Ideally, it'd be nice to be able to swap out the deeply buried `Transaction` class for a specific VM (`Chain -> VM -> VMState -> Block -> Transaction`) with a simple one-liner API.

```python
MainnetChain.new_magic_configuration_method(
    'frontier.vm.vm_state.block.transaction_class',
    MySpecialTransactionClass,
)
```

This would allow this sort of *deep* modification of the configuration such as injecting a *mixin* class with an overridden `apply_computation` method into *all* of the `Computation` classes.  At that point, it'd be easy for the `Computation` class to expose the functionality you suggest in the `post_opcode_callback`.

**Comment by NIC619** [comment]
@pipermerriam You are right. Altering the gas limit might end up in unexpected behaviors. And the way around it is instead checking those conditions(`PAYGAS` triggered or not, `OutOfGas` of not, etc.) after each opcode is executed as you suggested.

@jannikluhn I really like the `opcode_callback` idea. People can customize their own debug/analysis tool with this.
We can set injecting check as short term goal and opcode callback as long term goal(if `new_magic_configuration_method` idea is implemented).
